casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala

```diff
@@ -28,12 +27,17 @@ object CasperPacketHandler {
           message => EngineCell[F].read >>= (_.handle(peer, message))
         )
 
-  def fairDispatcher[F[_]: Concurrent: EngineCell: Running.RequestedBlocks: Log](
+  def fairDispatcher[F[_]: Concurrent: EngineCell: Running.RequestedBlocks: Log: Span: Metrics](
       maxPeerQueueSize: Int,
       giveUpAfterSkipped: Int,
       dropPeerAfterRetries: Int
-  ): F[PacketHandler[F]] = {
+  )(implicit spanF: Span[F]): F[PacketHandler[F]] = {
     import FairRoundRobinDispatcher._
+
+    implicit val fairDispatcherMetricsSource: Source =
+      Metrics.Source(Metrics.BaseSource, "CasperPacketHandler")
+
+    val fairLock                                      = MetricsSemaphore.single[F]
```

Comment:
I move the lock inside to not block every message which helps a little bit but it would still block in the `Handle` case and it is fine in `Pass` and `Drop` case.

Reply:
I mean the problem in the casper should be considered. We can split pr to solve them independently. I think this lock would become a huge problem when the network have many nodes.